cpp/src/join/join_common_utils.cuh

```diff
@@ -0,0 +1,263 @@
+                                              thrust::identity<size_type>()) -
+                              right_indices_complement->begin();
+    right_indices_complement->resize(indices_count, stream);
+  }
+
+  auto left_invalid_indices =
+    std::make_unique<rmm::device_uvector<size_type>>(right_indices_complement->size(), stream);
+  thrust::fill(rmm::exec_policy(stream),
+               left_invalid_indices->begin(),
+               left_invalid_indices->end(),
+               JoinNoneValue);
+
+  return std::make_pair(std::move(left_invalid_indices), std::move(right_indices_complement));
+}
+
+/**
+ * @brief Adds a pair of indices to the shared memory cache
+ *
+ * @param[in] first The first index in the pair
+ * @param[in] second The second index in the pair
+ * @param[in,out] current_idx_shared Pointer to shared index that determines
+ * where in the shared memory cache the pair will be written
+ * @param[in] warp_id The ID of the warp of the calling the thread
+ * @param[out] joined_shared_l Pointer to the shared memory cache for left indices
+ * @param[out] joined_shared_r Pointer to the shared memory cache for right indices
+ */
+__inline__ __device__ void add_pair_to_cache(const size_type first,
+                                             const size_type second,
+                                             size_type* current_idx_shared,
+                                             const int warp_id,
+                                             size_type* joined_shared_l,
+                                             size_type* joined_shared_r)
+{
+  size_type my_current_idx{atomicAdd(current_idx_shared + warp_id, size_type(1))};
+
+  // its guaranteed to fit into the shared cache
+  joined_shared_l[my_current_idx] = first;
+  joined_shared_r[my_current_idx] = second;
+}
+
+template <int num_warps, cudf::size_type output_cache_size>
+__device__ void flush_output_cache(const unsigned int activemask,
+                                   const cudf::size_type max_size,
+                                   const int warp_id,
+                                   const int lane_id,
+                                   cudf::size_type* current_idx,
+                                   cudf::size_type current_idx_shared[num_warps],
+                                   size_type join_shared_l[num_warps][output_cache_size],
+                                   size_type join_shared_r[num_warps][output_cache_size],
+                                   size_type* join_output_l,
+                                   size_type* join_output_r)
+{
+  // count how many active threads participating here which could be less than warp_size
+  int num_threads               = __popc(activemask);
+  cudf::size_type output_offset = 0;
+
+  if (0 == lane_id) { output_offset = atomicAdd(current_idx, current_idx_shared[warp_id]); }
+
+  output_offset = cub::ShuffleIndex<detail::warp_size>(output_offset, 0, activemask);
```

Comment:
Do you need to sync the warp before this due to the preceding divergent branch?

Reply:
I think a `__shfl` (or `__shfl_sync`) requires all threads in the warp (or at least, all threads marked as active in the `activemask`) to reach that point of execution, and will implicitly synchronize as needed to guarantee correctness. That also appears to be what's suggested [here](https://forums.developer.nvidia.com/t/shfl-sync-requires-syncwarp-for-every-call/58511).

Reply:
[It looks like cub may inline ptx and rely on cooperative groups instead of using `__shfl`](https://github.com/NVIDIA/cub/blob/36c7b552ecd785b1ef25061752d466957fce9e7c/cub/util_ptx.cuh#L337). However, this appears to still be safe since if I'm reading the documentation correctly [the warp-level sync is built into shfl primitives at the ptx level](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-shfl-sync).

Reply:
(Discussed offline) I'm satisfied. Though I suggest adding a comment specifying the expectations for active_mask for this function.

Reply:
Done.